Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPoolUnified.hs

```diff
@@ -0,0 +1,465 @@
+import Storage.Beam.Yudhishthira ()
+import qualified Storage.Cac.DriverIntelligentPoolConfig as CDIP
+import qualified Storage.Cac.TransporterConfig as SCTC
+import qualified Storage.CachedQueries.Driver.GoHomeRequest as CQDGR
+import qualified Storage.CachedQueries.ValueAddNP as CQVAN
+import Tools.Maps as Maps
+import Utils.Common.Cac.KeyNameConstants
+
+prepareDriverPoolBatch ::
+  ( EncFlow m r,
+    EsqDBReplicaFlow m r,
+    EsqDBFlow m r,
+    CacheFlow m r,
+    LT.HasLocationService m r
+  ) =>
+  [DVST.VehicleServiceTier] ->
+  DM.Merchant ->
+  DriverPoolConfig ->
+  DSR.SearchRequest ->
+  DST.SearchTry ->
+  [TripQuoteDetail] ->
+  PoolBatchNum ->
+  GoHomeConfig ->
+  m DriverPoolWithActualDistResultWithFlags
+prepareDriverPoolBatch cityServiceTiers merchant driverPoolCfg searchReq searchTry tripQuoteDetails startingbatchNum goHomeConfig = withLogTag ("startingbatchNum- (" <> show startingbatchNum <> ")") $ do
+  isValueAddNP <- CQVAN.isValueAddNP searchReq.bapId
+  previousBatchesDrivers <- getPreviousBatchesDrivers Nothing
+  previousBatchesDriversOnRide <- getPreviousBatchesDrivers (Just True)
+  let merchantOpCityId = searchReq.merchantOperatingCityId
+  logDebug $ "PreviousBatchesDrivers-" <> show previousBatchesDrivers
+  SDP.PrepareDriverPoolBatchEntity {..} <- prepareDriverPoolBatch' previousBatchesDrivers startingbatchNum merchantOpCityId searchReq.transactionId isValueAddNP
+  let finalPool = currentDriverPoolBatch <> currentDriverPoolBatchOnRide
+  SDP.incrementDriverRequestCount finalPool searchTry.id
+  pure $ buildDriverPoolWithActualDistResultWithFlags finalPool poolType nextScheduleTime (previousBatchesDrivers <> previousBatchesDriversOnRide)
+  where
+    buildDriverPoolWithActualDistResultWithFlags finalPool poolType nextScheduleTime prevBatchDrivers =
+      DriverPoolWithActualDistResultWithFlags
+        { driverPoolWithActualDistResult = finalPool,
+          poolType = poolType,
+          prevBatchDrivers = prevBatchDrivers,
+          nextScheduleTime = nextScheduleTime
+        }
+    getPreviousBatchesDrivers ::
+      ( EncFlow m r,
+        EsqDBReplicaFlow m r,
+        EsqDBFlow m r,
+        CacheFlow m r,
+        LT.HasLocationService m r
+      ) =>
+      Maybe Bool ->
+      m [Id Driver]
+    getPreviousBatchesDrivers mbOnRide = do
+      batches <- SDP.previouslyAttemptedDrivers searchTry.id mbOnRide
+      return $ (.driverPoolResult.driverId) <$> batches
+
+    prepareDriverPoolBatch' previousBatchesDrivers batchNum merchantOpCityId txnId isValueAddNP = withLogTag ("BatchNum - " <> show batchNum) $ do
+      radiusStep <- SDP.getPoolRadiusStep searchTry.id
+      transporterConfig <- SCTC.findByMerchantOpCityId merchantOpCityId Nothing >>= fromMaybeM (TransporterConfigDoesNotExist merchantOpCityId.getId)
+      intelligentPoolConfig <- CDIP.findByMerchantOpCityId merchantOpCityId (Just ((TransactionId . Id) txnId)) >>= fromMaybeM (InternalError "Intelligent Pool Config not found")
```

Comment:
Fetch when required.

Reply:
But we can think of optimisations also now only.

Reply:
Yup. Lets first fully move to tagged pooling and then think of it.